File: src/koreanpulse/billing/lemonsqueezy.py
"""Lemon Squeezy webhook handling — signature verification + event dispatch.

Lemon Squeezy is the Merchant of Record we use for global billing (handles
VAT, sales tax, GDPR data processing). They POST a JSON payload to our
webhook endpoint with an `X-Signature` header containing an HMAC-SHA256 of
the raw body computed with our `LEMONSQUEEZY_WEBHOOK_SECRET`.

Why we picked LS over Stripe for v0:
  - MoR — they handle Korean tax filing for us (avoids 사업자등록 + 부가세
    headaches in the early stage).
  - 5% + processing — predictable, no 1099-K, no Stripe Atlas trap.
  - Korean indie founders can sign up directly without US LLC.

Variant ID → Plan mapping is loaded from env so the same code works in
test/live storefronts. See `.env.example` for `LEMONSQUEEZY_VARIANT_*`.

Reference: <https://docs.lemonsqueezy.com/help/webhooks>
"""
from __future__ import annotations

import hashlib
import hmac
import logging
import os
from datetime import datetime, timezone
from typing import Any, Optional

from koreanpulse.license import (
    LIFETIME_DEAL_PLAN,
    License,
    LicenseStore,
    Plan,
    aget_default_store,
    is_lifetime,
    issue_license_key,
)

logger = logging.getLogger(__name__)


# ── Variant ID → Plan ──────────────────────────────────────────────────────
# In Lemon Squeezy you create one Variant per pricing tier.
# Register their numeric IDs here via env so the same handler works in
# test mode and live mode without code changes.
#
# Pricing v2 (2026-05-05): Workflow-priced 3-tier ladder. Required in
# production: SOLO + ANALYST + DESK. Deprecated slots (PRO / STARTER /
# INDIE / ENTERPRISE) remain wired for back-compat with historical
# storefronts — leave them unset in production.
def _read_variant_map() -> dict[str, Plan]:
    """Build the variant_id (str) → Plan map from env vars.

    Production requires SOLO + ANALYST + DESK. Deprecated tiers (PRO,
    STARTER, INDIE, ENTERPRISE) are kept as no-op slots for back-compat
    with historical storefronts; leave them unset and the variant map
    only contains the active 3-tier ladder.
    """
    out: dict[str, Plan] = {}
    pairs = [
        ("LEMONSQUEEZY_VARIANT_SOLO", Plan.SOLO),
        ("LEMONSQUEEZY_VARIANT_ANALYST", Plan.ANALYST),
        ("LEMONSQUEEZY_VARIANT_DESK", Plan.DESK),
        # Deprecated — historical storefront rows. Unused in production.
        ("LEMONSQUEEZY_VARIANT_PRO", Plan.PRO),
        ("LEMONSQUEEZY_VARIANT_STARTER", Plan.STARTER),
        ("LEMONSQUEEZY_VARIANT_INDIE", Plan.INDIE),
        ("LEMONSQUEEZY_VARIANT_ENTERPRISE", Plan.ENTERPRISE),
    ]
    for env_name, plan in pairs:
        vid = os.environ.get(env_name, "").strip()
        if vid:
            out[vid] = plan
    return out


def _read_lifetime_variant() -> Optional[str]:
    return os.environ.get("LEMONSQUEEZY_VARIANT_LIFETIME", "").strip() or None


# ── Signature verification ─────────────────────────────────────────────────


class WebhookVerificationError(Exception):
    """Raised when the X-Signature header doesn't match the body HMAC."""


def verify_signature(*, body: bytes, signature_header: str, secret: str) -> bool:
    """Return True iff `signature_header` matches HMAC-SHA256(body, secret).

    Constant-time compare to defeat timing attacks.
    """
    if not signature_header or not secret:
        return False
    digest = hmac.new(
        secret.encode("utf-8"), body, hashlib.sha256
    ).hexdigest()
    return hmac.compare_digest(digest, signature_header.strip())


# ── Idempotency ────────────────────────────────────────────────────────────
# Lemon Squeezy retries webhooks on non-2xx responses. We dedupe by
# `meta.webhook_id` to keep handler effects exactly-once.


class _SeenEvents:
    """Bounded LRU of webhook IDs we've already processed.

    In-memory only — fine until we go multi-instance. Swap to Postgres /
    Redis once we have horizontal replicas.
    """

    def __init__(self, max_size: int = 10_000) -> None:
        self._set: set[str] = set()
        self._order: list[str] = []
        self._max = max_size

    def is_new(self, webhook_id: str) -> bool:
        if webhook_id in self._set:
            return False
        self._set.add(webhook_id)
        self._order.append(webhook_id)
        if len(self._order) > self._max:
            stale = self._order.pop(0)
            self._set.discard(stale)
        return True


_seen = _SeenEvents()


# ── Event dispatch ─────────────────────────────────────────────────────────


class HandlerResult:
    """Outcome of processing one webhook event."""

    def __init__(self, *, ok: bool, action: str, license_key: Optional[str] = None,
                 message: str = "") -> None:
        self.ok = ok
        self.action = action
        self.license_key = license_key
        self.message = message

    def to_dict(self) -> dict[str, Any]:
        return {
            "ok": self.ok,
            "action": self.action,
            "license_key": self.license_key,
            "message": self.message,
        }


def _extract_email(attributes: dict) -> str:
    return (
        attributes.get("user_email")
        or attributes.get("email")
        or attributes.get("customer_email")
        or ""
    ).strip().lower()


def _extract_variant_id(attributes: dict) -> Optional[str]:
    """Pull the variant ID from a subscription or order payload."""
    if "variant_id" in attributes:
        return str(attributes["variant_id"])
    # Order payload: variant lives under first_order_item
    first = attributes.get("first_order_item") or {}
    if isinstance(first, dict) and "variant_id" in first:
        return str(first["variant_id"])
    return None


# Audience composition signal — captured at checkout via a Lemon Squeezy
# custom field labelled "role". Allowed values match landing/app/api/notify.
_ALLOWED_ROLES = {
    "analyst", "rotator", "diaspora", "journalist", "developer", "other",
}


def _extract_self_description(payload: dict, attributes: dict) -> Optional[str]:
    """Best-effort role/self-description extraction.

    Lemon Squeezy surfaces checkout custom fields in different shapes
    depending on whether the merchant defined them on the variant
    (`first_order_item.product_options.custom`) or on the storefront
    checkout (`meta.custom_data.role`). Both paths are tried; the first
    valid match wins. Unknown values normalise to "other".
    """
    candidates: list = []

    meta = payload.get("meta") or {}
    cd = meta.get("custom_data") or {}
    if isinstance(cd, dict):
        candidates.append(cd.get("role"))

    first = attributes.get("first_order_item") or {}
    if isinstance(first, dict):
        po = first.get("product_options") or {}
        if isinstance(po, dict):
            custom = po.get("custom") or {}
            if isinstance(custom, dict):
                candidates.append(custom.get("role"))

    cf = attributes.get("custom_fields_responses") or {}
    if isinstance(cf, dict):
        candidates.append(cf.get("role"))

    for c in candidates:
        if not c:
            continue
        s = str(c).strip().lower()
        if not s:
            continue
        return s if s in _ALLOWED_ROLES else "other"
    return None


async def handle_event(
    payload: dict,
    *,
    store: Optional[LicenseStore] = None,
) -> HandlerResult:
    """Dispatch one Lemon Squeezy event payload.

    Idempotent on `meta.webhook_id`. Safe to call from any FastAPI route
    after signature verification.

    Returns a HandlerResult; the route maps it to an HTTP status (always
    200 unless we want LS to retry).
    """
    store = store or await aget_default_store()

    meta = payload.get("meta", {}) or {}
    event_name = (meta.get("event_name") or "").strip()
    webhook_id = (meta.get("webhook_id") or "").strip()

    if webhook_id and not _seen.is_new(webhook_id):
        return HandlerResult(
            ok=True, action="duplicate_ignored",
            message=f"already processed webhook_id={webhook_id}",
        )

    data = payload.get("data") or {}
    attributes = data.get("attributes") or {}
    # Lemon Squeezy puts the canonical entity ID at `data.id`, not in
    # attributes. For subscription events that's the subscription_id; for
    # order_created it's the order_id. Pull it once here.
    entity_id = str(data.get("id") or "")
    email = _extract_email(attributes)
    variant_id = _extract_variant_id(attributes)
    self_description = _extract_self_description(payload, attributes)

    if event_name in ("subscription_created", "subscription_resumed"):
        return await _on_subscription_active(
            store=store, email=email, variant_id=variant_id,
            attributes=attributes, entity_id=entity_id,
            self_description=self_description,
        )

    if event_name == "subscription_updated":
        return await _on_subscription_updated(
            store=store, email=email, variant_id=variant_id, attributes=attributes,
        )

    if event_name in ("subscription_cancelled", "subscription_expired"):
        return await _on_subscription_inactive(
            store=store, email=email, attributes=attributes,
        )

    if event_name == "subscription_payment_success":
        return await _on_payment_success(
            store=store, email=email, attributes=attributes,
        )

    if event_name == "subscription_payment_failed":
        return await _on_payment_failed(
            store=store, email=email, attributes=attributes,
        )

    if event_name == "order_created":
        return await _on_order_created(
            store=store, email=email, variant_id=variant_id,
            attributes=attributes, entity_id=entity_id,
            self_description=self_description,
        )

    return HandlerResult(
        ok=True, action="ignored",
        message=f"no handler for event_name={event_name!r}",
    )


# ── Per-event handlers ─────────────────────────────────────────────────────


async def _on_subscription_active(
    *, store: LicenseStore, email: str, variant_id: Optional[str],
    attributes: dict, entity_id: str,
    self_description: Optional[str] = None,
) -> HandlerResult:
    if not email:
        return HandlerResult(ok=False, action="error", message="missing email")
    plan = _read_variant_map().get(variant_id or "")
    if plan is None:
        return HandlerResult(
            ok=False, action="error",
            message=f"unknown variant_id={variant_id!r}; configure LEMONSQUEEZY_VARIANT_* env",
        )

    # If this email already has an active license, upgrade it in place.
    existing = await store.find_by_email(email)
    if existing is not None:
        existing.plan = plan
        existing.active = True
        update: dict = {
            "ls_subscription_id": entity_id,
            "ls_variant_id": variant_id,
            "ls_status": attributes.get("status"),
        }
        # Only overwrite self_description on a fresh value — don't blank
        # out a previously-captured role on later subscription updates.
        if self_description:
            update["self_description"] = self_description
        existing.metadata.update(update)
        await store.save(existing)
        return HandlerResult(
            ok=True, action="upgraded", license_key=existing.key,
            message=f"existing license upgraded to {plan.value}",
        )

    metadata: dict = {
        "ls_subscription_id": entity_id,
        "ls_variant_id": variant_id,
        "ls_status": attributes.get("status"),
        "issued_via": "subscription_created",
    }
    if self_description:
        metadata["self_description"] = self_description

    lic = License(
        key=issue_license_key(),
        plan=plan,
        customer_email=email,
        active=True,
        metadata=metadata,
    )
    await store.save(lic)
    return HandlerResult(
        ok=True, action="issued", license_key=lic.key,
        message=f"new {plan.value} license issued for {email}",
    )


async def _on_subscription_updated(
    *, store: LicenseStore, email: str, variant_id: Optional[str], attributes: dict,
) -> HandlerResult:
    if not email:
        return HandlerResult(ok=False, action="error", message="missing email")
    lic = await store.find_by_email(email)
    if lic is None:
        # Treat as new — but we don't have the original data.id here, so
        # leave entity_id empty. subscription_created should have caught it.
        return await _on_subscription_active(
            store=store, email=email, variant_id=variant_id,
            attributes=attributes, entity_id="",
        )

    new_plan = _read_variant_map().get(variant_id or "") if variant_id else None
    if new_plan is not None and new_plan != lic.plan:
        lic.plan = new_plan
    status = attributes.get("status", "")
    lic.active = status in ("active", "on_trial", "past_due")  # past_due still serves
    lic.metadata["ls_status"] = status
    await store.save(lic)
    return HandlerResult(
        ok=True, action="updated", license_key=lic.key,
        message=f"license updated plan={lic.plan.value} active={lic.active}",
    )


async def _on_subscription_inactive(
    *, store: LicenseStore, email: str, attributes: dict,
) -> HandlerResult:
    lic = await store.find_by_email(email)
    if lic is None:
        return HandlerResult(ok=True, action="noop", message="no license to deactivate")
    # Lifetime deal customers are never deactivated by subscription events.
    if is_lifetime(lic):
        return HandlerResult(
            ok=True, action="noop",
            license_key=lic.key, message="lifetime license preserved",
        )
    lic.active = False
    lic.metadata["ls_status"] = attributes.get("status", "cancelled")
    await store.save(lic)
    return HandlerResult(
        ok=True, action="deactivated", license_key=lic.key,
        message=f"license deactivated for {email}",
    )


async def _on_payment_success(
    *, store: LicenseStore, email: str, attributes: dict,
) -> HandlerResult:
    """Reset the rolling period counter on successful renewal."""
    lic = await store.find_by_email(email)
    if lic is None:
        return HandlerResult(ok=True, action="noop", message="no license to reset")
    lic.period_calls = 0
    lic.period_started_at = datetime.now(timezone.utc)
    lic.active = True
    await store.save(lic)
    return HandlerResult(
        ok=True, action="period_reset", license_key=lic.key,
        message="usage counter reset on renewal",
    )


async def _on_payment_failed(
    *, store: LicenseStore, email: str, attributes: dict,
) -> HandlerResult:
    """Lemon Squeezy will retry; we keep the license active during grace."""
    lic = await store.find_by_email(email)
    if lic is None:
        return HandlerResult(ok=True, action="noop")
    lic.metadata["ls_last_payment_failed_at"] = datetime.now(timezone.utc).isoformat()
    await store.save(lic)
    return HandlerResult(
        ok=True, action="payment_failed_logged", license_key=lic.key,
        message="payment failure logged; license still active during grace period",
    )


async def _on_order_created(
    *, store: LicenseStore, email: str, variant_id: Optional[str],
    attributes: dict, entity_id: str,
    self_description: Optional[str] = None,
) -> HandlerResult:
    """Handle one-time orders (used for the lifetime deal)."""
    if not email:
        return HandlerResult(ok=False, action="error", message="missing email")
    lifetime_variant = _read_lifetime_variant()
    if not lifetime_variant or variant_id != lifetime_variant:
        return HandlerResult(
            ok=True, action="ignored",
            message=f"order_created for variant={variant_id!r} is not the lifetime SKU",
        )

    existing = await store.find_by_email(email)
    if existing is not None and is_lifetime(existing):
        return HandlerResult(
            ok=True, action="duplicate_lifetime",
            license_key=existing.key, message="email already has lifetime license",
        )

    # Auto-assign next deal_seq based on current count of lifetime licenses.
    deal_seq = await store.next_lifetime_seq()
    metadata: dict = {
        "lifetime": True,
        "deal_seq": deal_seq,
        "ls_order_id": entity_id,
        "ls_variant_id": variant_id,
        "issued_via": "order_created",
    }
    if self_description:
        metadata["self_description"] = self_description

    lic = License(
        key=issue_license_key(),
        plan=LIFETIME_DEAL_PLAN,
        customer_email=email,
        active=True,
        metadata=metadata,
    )
    await store.save(lic)
    return HandlerResult(
        ok=True, action="lifetime_issued", license_key=lic.key,
        message=f"lifetime deal #{deal_seq} issued to {email}",
    )

File: webhook-worker/src/lemonsqueezy.ts
/**
 * Lemon Squeezy webhook handling for the simplified pricing model.
 * TypeScript port of `koreanpulse.billing.lemonsqueezy` over Cloudflare D1.
 *
 * Flow per webhook:
 *   1. HMAC-SHA256 signature verify (LS_WEBHOOK_SECRET).
 *   2. JSON parse.
 *   3. Idempotency check via D1 `webhook_events.webhook_id` PK.
 *   4. Dispatch on `meta.event_name` to per-event handler.
 *   5. Handler upserts to D1 `licenses`.
 *   6. Always return 200 — surface failure in body so the LS dashboard
 *      shows it but no retry storm.
 */

import type { D1Database } from "@cloudflare/workers-types";
import {
  findByEmail,
  getByKey,
  issueLicenseKey,
  markEventSeen,
  nextLifetimeSeq,
  upsertLicense,
  type License,
  type Plan,
} from "./license";

export interface Env {
  DB: D1Database;
  LEMONSQUEEZY_WEBHOOK_SECRET: string;
  KOREANPULSE_CACHE_SHARED_SECRET: string;
  // Active tiers (pricing v2)
  LEMONSQUEEZY_VARIANT_SOLO: string;
  LEMONSQUEEZY_VARIANT_ANALYST: string;
  LEMONSQUEEZY_VARIANT_DESK: string;
  // Lifetime: design-partner-only, 20 seats max, NOT in public pricing.
  LEMONSQUEEZY_VARIANT_LIFETIME: string;
  // Deprecated / back-compat — historical storefront rows. Empty in prod.
  LEMONSQUEEZY_VARIANT_PRO?: string;
  LEMONSQUEEZY_VARIANT_STARTER?: string;
  LEMONSQUEEZY_VARIANT_INDIE?: string;
  LEMONSQUEEZY_VARIANT_ENTERPRISE?: string;
  VALIDATE_CACHE_TTL_SECONDS?: string;
}

export interface HandlerResult {
  ok: boolean;
  action: string;
  license_key?: string;
  message?: string;
}

const ALLOWED_ROLES = new Set([
  "analyst", "rotator", "diaspora", "journalist", "developer", "other",
]);

function readVariantMap(env: Env): Map<string, Plan> {
  const m = new Map<string, Plan>();
  const pairs: [string | undefined, Plan][] = [
    // Active 2026-05-05+ (workflow-priced 3-tier ladder)
    [env.LEMONSQUEEZY_VARIANT_SOLO, "solo"],
    [env.LEMONSQUEEZY_VARIANT_ANALYST, "analyst"],
    [env.LEMONSQUEEZY_VARIANT_DESK, "desk"],
    // Deprecated — historical storefront rows. Normally empty in production.
    [env.LEMONSQUEEZY_VARIANT_PRO, "pro"],
    [env.LEMONSQUEEZY_VARIANT_STARTER, "starter"],
    [env.LEMONSQUEEZY_VARIANT_INDIE, "indie"],
    [env.LEMONSQUEEZY_VARIANT_ENTERPRISE, "enterprise"],
  ];
  for (const [vid, plan] of pairs) {
    const trimmed = (vid ?? "").trim();
    if (trimmed) m.set(trimmed, plan);
  }
  return m;
}

// ── Signature verification (HMAC-SHA256, constant-time compare) ────────────

export async function verifyLsSignature(
  body: string,
  signatureHeader: string,
  secret: string,
): Promise<boolean> {
  if (!signatureHeader || !secret) return false;
  const expected = await hmacSha256Hex(secret, body);
  return constantTimeEq(expected, signatureHeader.trim());
}

async function hmacSha256Hex(secret: string, message: string): Promise<string> {
  const key = await crypto.subtle.importKey(
    "raw",
    new TextEncoder().encode(secret),
    { name: "HMAC", hash: "SHA-256" },
    false,
    ["sign"],
  );
  const sig = await crypto.subtle.sign(
    "HMAC",
    key,
    new TextEncoder().encode(message),
  );
  return Array.from(new Uint8Array(sig))
    .map((b) => b.toString(16).padStart(2, "0"))
    .join("");
}

function constantTimeEq(a: string, b: string): boolean {
  if (a.length !== b.length) return false;
  let diff = 0;
  for (let i = 0; i < a.length; i++) {
    diff |= a.charCodeAt(i) ^ b.charCodeAt(i);
  }
  return diff === 0;
}

// ── Field extractors ───────────────────────────────────────────────────────

function extractEmail(attrs: Record<string, unknown>): string {
  const v = (attrs.user_email ?? attrs.email ?? attrs.customer_email ?? "") as string;
  return String(v).trim().toLowerCase();
}

function extractVariantId(attrs: Record<string, unknown>): string | undefined {
  if ("variant_id" in attrs) return String(attrs.variant_id);
  const first = (attrs.first_order_item ?? {}) as Record<string, unknown>;
  if (first && typeof first === "object" && "variant_id" in first) {
    return String(first.variant_id);
  }
  return undefined;
}

function extractSelfDescription(
  payload: Record<string, unknown>,
  attrs: Record<string, unknown>,
): string | undefined {
  const candidates: unknown[] = [];

  const meta = (payload.meta ?? {}) as Record<string, unknown>;
  const cd = (meta.custom_data ?? {}) as Record<string, unknown>;
  if (cd && typeof cd === "object") candidates.push((cd as Record<string, unknown>).role);

  const first = (attrs.first_order_item ?? {}) as Record<string, unknown>;
  if (first && typeof first === "object") {
    const po = (first.product_options ?? {}) as Record<string, unknown>;
    if (po && typeof po === "object") {
      const custom = (po.custom ?? {}) as Record<string, unknown>;
      if (custom && typeof custom === "object") {
        candidates.push((custom as Record<string, unknown>).role);
      }
    }
  }

  const cf = (attrs.custom_fields_responses ?? {}) as Record<string, unknown>;
  if (cf && typeof cf === "object") candidates.push((cf as Record<string, unknown>).role);

  for (const c of candidates) {
    if (!c) continue;
    const s = String(c).trim().toLowerCase();
    if (!s) continue;
    return ALLOWED_ROLES.has(s) ? s : "other";
  }
  return undefined;
}

// ── Event dispatch ─────────────────────────────────────────────────────────

export async function handleEvent(env: Env, payload: any): Promise<HandlerResult> {
  const meta = (payload?.meta ?? {}) as Record<string, unknown>;
  const eventName = String(meta.event_name ?? "").trim();
  const webhookId = String(meta.webhook_id ?? "").trim();

  const data = (payload?.data ?? {}) as Record<string, unknown>;
  const attrs = (data.attributes ?? {}) as Record<string, unknown>;
  const entityId = String(data.id ?? "");
  const email = extractEmail(attrs);
  const variantId = extractVariantId(attrs);
  const selfDescription = extractSelfDescription(payload, attrs);

  // Idempotency. Note: the row gets written ONLY after the dispatch
  // succeeds (so a transient failure doesn't lose retries). We pre-check
  // here and write the audit row at the end of dispatch.
  if (webhookId) {
    const existing = await env.DB
      .prepare("SELECT 1 FROM webhook_events WHERE webhook_id = ?")
      .bind(webhookId)
      .first();
    if (existing) {
      return { ok: true, action: "duplicate_ignored", message: `already processed ${webhookId}` };
    }
  }

  let result: HandlerResult;

  switch (eventName) {
    case "subscription_created":
    case "subscription_resumed":
      result = await onSubscriptionActive(env, { email, variantId, attrs, entityId, selfDescription });
      break;
    case "subscription_updated":
      result = await onSubscriptionUpdated(env, { email, variantId, attrs });
      break;
    case "subscription_cancelled":
    case "subscription_expired":
      result = await onSubscriptionInactive(env, { email, attrs });
      break;
    case "subscription_payment_success":
      result = await onPaymentSuccess(env, { email });
      break;
    case "subscription_payment_failed":
      result = await onPaymentFailed(env, { email });
      break;
    case "order_created":
      result = await onOrderCreated(env, { email, variantId, entityId, selfDescription });
      break;
    default:
      result = { ok: true, action: "ignored", message: `no handler for event_name=${eventName}` };
  }

  // Persist idempotency log AFTER dispatch finishes — only successful
  // events are recorded so retries can re-attempt failed ones.
  if (webhookId && result.ok) {
    await markEventSeen(env.DB, webhookId, eventName, {
      license_key: result.license_key,
      action: result.action,
      note: result.message,
    });
  }

  return result;
}

// ── Per-event handlers ─────────────────────────────────────────────────────

interface SubArgs {
  email: string;
  variantId?: string;
  attrs: Record<string, unknown>;
  entityId: string;
  selfDescription?: string;
}

async function onSubscriptionActive(env: Env, a: SubArgs): Promise<HandlerResult> {
  if (!a.email) return { ok: false, action: "error", message: "missing email" };
  const plan = readVariantMap(env).get(a.variantId ?? "");
  if (!plan) {
    return {
      ok: false, action: "error",
      message: `unknown variant_id=${a.variantId}; configure LEMONSQUEEZY_VARIANT_*`,
    };
  }

  const existing = await findByEmail(env.DB, a.email);
  const status = String(a.attrs.status ?? "");
  const subId = a.entityId;

  if (existing) {
    existing.plan = plan;
    existing.active = true;
    existing.metadata = {
      ...existing.metadata,
      ls_subscription_id: subId,
      ls_variant_id: a.variantId,
      ls_status: status,
    };
    if (a.selfDescription) {
      existing.metadata.self_description = a.selfDescription;
    }
    await upsertLicense(env.DB, existing);
    return {
      ok: true, action: "upgraded", license_key: existing.key,
      message: `existing license upgraded to ${plan}`,
    };
  }

  const now = new Date().toISOString();
  const lic: License = {
    key: issueLicenseKey(),
    plan,
    customer_email: a.email,
    active: true,
    created_at: now,
    period_calls: 0,
    period_started_at: now,
    metadata: {
      ls_subscription_id: subId,
      ls_variant_id: a.variantId,
      ls_status: status,
      issued_via: "subscription_created",
      ...(a.selfDescription ? { self_description: a.selfDescription } : {}),
    },
    is_lifetime: false,
    deal_seq: null,
    updated_at: now,
  };
  await upsertLicense(env.DB, lic);
  return {
    ok: true, action: "issued", license_key: lic.key,
    message: `new ${plan} license issued for ${a.email}`,
  };
}

async function onSubscriptionUpdated(
  env: Env,
  a: { email: string; variantId?: string; attrs: Record<string, unknown> },
): Promise<HandlerResult> {
  if (!a.email) return { ok: false, action: "error", message: "missing email" };
  const lic = await findByEmail(env.DB, a.email);
  if (!lic) {
    // Treat as new but without entity_id (we don't have it for "updated"
    // events that arrive for never-seen subscriptions — rare).
    return await onSubscriptionActive(env, { email: a.email, variantId: a.variantId, attrs: a.attrs, entityId: "" });
  }
  const newPlan = a.variantId ? readVariantMap(env).get(a.variantId) : undefined;
  if (newPlan && newPlan !== lic.plan) lic.plan = newPlan;
  const status = String(a.attrs.status ?? "");
  lic.active = status === "active" || status === "on_trial" || status === "past_due";
  lic.metadata = { ...lic.metadata, ls_status: status };
  await upsertLicense(env.DB, lic);
  return {
    ok: true, action: "updated", license_key: lic.key,
    message: `license updated plan=${lic.plan} active=${lic.active}`,
  };
}

async function onSubscriptionInactive(
  env: Env,
  a: { email: string; attrs: Record<string, unknown> },
): Promise<HandlerResult> {
  const lic = await findByEmail(env.DB, a.email);
  if (!lic) return { ok: true, action: "noop", message: "no license to deactivate" };
  if (lic.is_lifetime) {
    return {
      ok: true, action: "noop",
      license_key: lic.key, message: "lifetime license preserved",
    };
  }
  lic.active = false;
  lic.metadata = { ...lic.metadata, ls_status: String(a.attrs.status ?? "cancelled") };
  await upsertLicense(env.DB, lic);
  return {
    ok: true, action: "deactivated", license_key: lic.key,
    message: `license deactivated for ${a.email}`,
  };
}

async function onPaymentSuccess(env: Env, a: { email: string }): Promise<HandlerResult> {
  const lic = await findByEmail(env.DB, a.email);
  if (!lic) return { ok: true, action: "noop", message: "no license to reset" };
  lic.period_calls = 0;
  lic.period_started_at = new Date().toISOString();
  lic.active = true;
  await upsertLicense(env.DB, lic);
  return {
    ok: true, action: "period_reset", license_key: lic.key,
    message: "usage counter reset on renewal",
  };
}

async function onPaymentFailed(env: Env, a: { email: string }): Promise<HandlerResult> {
  const lic = await findByEmail(env.DB, a.email);
  if (!lic) return { ok: true, action: "noop" };
  lic.metadata = { ...lic.metadata, ls_last_payment_failed_at: new Date().toISOString() };
  await upsertLicense(env.DB, lic);
  return {
    ok: true, action: "payment_failed_logged", license_key: lic.key,
    message: "payment failure logged; license still active during grace period",
  };
}

async function onOrderCreated(
  env: Env,
  a: { email: string; variantId?: string; entityId: string; selfDescription?: string },
): Promise<HandlerResult> {
  if (!a.email) return { ok: false, action: "error", message: "missing email" };
  const lifetimeVariant = (env.LEMONSQUEEZY_VARIANT_LIFETIME ?? "").trim();
  if (!lifetimeVariant || a.variantId !== lifetimeVariant) {
    return {
      ok: true, action: "ignored",
      message: `order_created variant=${a.variantId} is not the lifetime SKU`,
    };
  }

  const existing = await findByEmail(env.DB, a.email);
  if (existing && existing.is_lifetime) {
    return {
      ok: true, action: "duplicate_lifetime",
      license_key: existing.key, message: "email already has lifetime license",
    };
  }

  const dealSeq = await nextLifetimeSeq(env.DB);
  const now = new Date().toISOString();
  const lic: License = {
    key: issueLicenseKey(),
    plan: "analyst",  // pricing v2: design-partner lifetime grants Analyst forever
    customer_email: a.email,
    active: true,
    created_at: now,
    period_calls: 0,
    period_started_at: now,
    metadata: {
      lifetime: true,
      deal_seq: dealSeq,
      deal_price_usd: 299,
      ls_order_id: a.entityId,
      ls_variant_id: a.variantId,
      issued_via: "order_created",
      ...(a.selfDescription ? { self_description: a.selfDescription } : {}),
    },
    is_lifetime: true,
    deal_seq: dealSeq,
    updated_at: now,
  };
  await upsertLicense(env.DB, lic);
  return {
    ok: true, action: "lifetime_issued", license_key: lic.key,
    message: `lifetime deal #${dealSeq} issued to ${a.email}`,
  };
}

// Re-export for getByKey usage from index
export { getByKey };
